components/support/nimbus-fml/src/backends/kotlin/mod.rs

```diff
@@ -0,0 +1,162 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::Result;
+use askama::Template;
+
+use crate::intermediate_representation::FeatureManifest;
+use crate::{Config, GenerateStructCmd};
+
+mod gen_structs;
+
+pub(crate) fn generate_struct(
+    manifest: FeatureManifest,
+    config: Option<Config>,
+    cmd: GenerateStructCmd,
+) -> Result<()> {
+    let config = config.unwrap_or_default();
+    let kt = gen_structs::FeatureManifestDeclaration::new(config, &manifest);
+
+    let path = cmd.output;
+    let contents = kt.render()?;
+
+    std::fs::write(path, contents)?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+pub mod test {
+    use crate::error::FMLError;
+    use crate::util::{join, pkg_dir, sdk_dir};
+    use anyhow::{bail, Result};
+    use std::path::{Path, PathBuf};
+    use std::process::Command;
+    use tempdir::TempDir;
+
+    // The root of the Android kotlin package structure
+    fn sdk_android_dir() -> String {
+        join(sdk_dir(), "android/src/main/java")
+    }
+
+    // The directory with the mock implementations of Android
+    // used for testing.
+    fn runtime_dir() -> String {
+        join(pkg_dir(), "fixtures/android/runtime")
+    }
+
+    // We'll put our test scripts in here.
+    fn tests_dir() -> String {
+        join(pkg_dir(), "fixtures/android/tests")
+    }
+
+    // The jar archive we need to do JSON with in Kotlin/Java.
+    // This is the same library as bundled in Android.
+    fn json_jar() -> String {
+        join(runtime_dir(), "json.jar")
+    }
+
+    // The file with the kt implementation of FeatureVariables
+    fn variables_kt() -> String {
+        join(
+            sdk_android_dir(),
+            "org/mozilla/experiments/nimbus/FeatureVariables.kt",
+        )
+    }
+
+    fn nimbus_internals_kt() -> String {
+        join(sdk_android_dir(), "org/mozilla/experiments/nimbus/internal")
+    }
+
+    // The file with the kt implementation of FeatureVariables
+    fn features_kt() -> String {
+        join(
+            sdk_android_dir(),
+            "org/mozilla/experiments/nimbus/FeaturesInterface.kt",
+        )
+    }
+
+    fn classpath(classes: &Path) -> Result<String> {
+        Ok(format!("{}:{}", json_jar(), classes.to_str().unwrap()))
+    }
+
+    fn detect_kotlinc() -> Result<bool> {
+        let output = Command::new("which").arg("kotlinc").output()?;
```

Comment:
I'm trying to remember if our tests run in any non-unix environment (or any environment without the `which` command) but I think we should be okay!